src/System.Buffers/src/System/Buffers/DefaultArrayPoolBucket.cs

```diff
@@ -80,7 +96,17 @@ internal void Return(T[] buffer)
                 // we can just drop this buffer
                 if (_index != 0)
                 {
+                    // If we have an event listener, we are in "debug" mode, so do a safety check
```

Comment:
Events enabled isn't "debug" per se (though I get your point).  That would be more if this were built in Debug or if Debugger.IsAttached.  Often events get turned in production to diagnose something or see how something is performing, etc., and you want turning on events to be as lightweight as possible.  Turning on events also shouldn't alter behavior.  But here it looks like we're potentially doing a lot of work (maybe not?) and also throwing in places we weren't otherwise.  I'm not saying we shouldn't do this, but definitely think hard about doing so.